components/raftstore-v2/src/operation/command/admin/merge/commit.rs

```diff
@@ -0,0 +1,648 @@
+            } else {
+                None
+            };
+        }
+        let first = entries.first().unwrap();
+        // make sure message should be with index not smaller than committed
+        let mut log_idx = first.get_index() - 1;
+        debug!(
+            self.logger,
+            "append merge entries";
+            "log_index" => log_idx,
+            "merge_commit" => merge.get_commit(),
+            "commit_index" => self.raft_group().raft.raft_log.committed,
+        );
+        if log_idx < self.raft_group().raft.raft_log.committed {
+            // There may be some logs not included in CommitMergeRequest's entries, like
+            // CompactLog, so the commit index may exceed the last index of the entires from
+            // CommitMergeRequest. If that, no need to append
+            if self.raft_group().raft.raft_log.committed - log_idx >= entries.len() as u64 {
+                return None;
+            }
+            entries = &entries[(self.raft_group().raft.raft_log.committed - log_idx) as usize..];
+            log_idx = self.raft_group().raft.raft_log.committed;
+        }
+        let log_term = self.index_term(log_idx);
+
+        let last_log = entries.last().unwrap();
+        if last_log.term > self.term() {
+            // Hack: In normal flow, when leader sends the entries, it will use a term
+            // that's not less than the last log term. And follower will update its states
+            // correctly. For merge, we append the log without raft, so we have to take care
+            // of term explicitly to get correct metadata.
+            info!(
+                self.logger,
+                "become follower for new logs";
+                "new_log_term" => last_log.term,
+                "new_log_index" => last_log.index,
+                "term" => self.term(),
+            );
+            self.raft_group_mut()
+                .raft
+                .become_follower(last_log.term, INVALID_ID);
+        }
+
+        self.raft_group_mut()
+            .raft
+            .raft_log
+            .maybe_append(log_idx, log_term, merge.get_commit(), entries)
+            .map(|(_, last_index)| last_index)
+    }
+
+    // Called on source peer.
+    pub fn post_prepare_merge<T: Transport>(&mut self, store_ctx: &mut StoreContext<EK, ER, T>) {
+        assert!(self.applied_merge_state().is_some());
+        if let Some(c) = &self.merge_context().unwrap().catch_up_logs
+            && self.catch_up_logs_ready(c)
+        {
+            let c = self.merge_context_mut().catch_up_logs.take().unwrap();
+            if c.tx.send(()).is_err() {
+                error!(
+                    self.logger,
+                    "failed to respond to merge target, are we shutting down?"
+                );
+                // TODO: rollback.
+            }
+            self.take_merge_context();
+            self.mark_for_destroy(None);
+            return;
+        }
+        self.on_check_merge(store_ctx);
```

Comment:
This is necessary only when `catch_up_logs` is `None`.

Reply:
I don't think so. This is the only place that `AskCommitMerge` is generated. No one will generate a catch_up_logs if this is removed.

Reply:
`catch_up_logs` will be set after the sibling peer apply commit merge, which can be proposed by itself or the leader on other node.

If `catch_up_logs` is `Some`, then the source peer must be merged in the end and the target peer must have applied commit merge. So there is no point to send `AskCommitMerge` again.

Reply:
Oh, I misread your comment.